File: Backend/app/core/permissions.py
"""Centralised permission registry and authorization helpers.

This module is **additive** and **opt-in**. The existing
:func:`app.core.deps.role_required` dependency keeps working unchanged — every
router that uses it today continues to operate exactly as before. New routes
(and refactors of old ones, when desired) can switch to permission-based
authorization via :func:`require_permission` for a more future-proof API.

Rationale
---------
Scattered role checks like ``if user.role.name == "admin": ...`` are brittle:
adding a new role (for example ``IT_admin`` or ``HR``) requires hunting down
every comparison. A permission registry decouples *what* the user can do from
*which role* they have, so introducing a new role only requires updating one
table.

Backwards compatibility guarantees
----------------------------------
- Role names (``admin``, ``manager``, ``employee``) are preserved verbatim.
- All current ``role_required(...)`` call sites continue to work.
- Permission checks fall through to the legacy role check, so a JWT minted
  by the legacy ``/api/v1/auth/login`` (which carries ``role`` as a claim)
  is honoured even if the DB row lookup is bypassed.
- No new DB tables, columns, or migrations.
"""
from __future__ import annotations

from typing import Dict, Iterable, Set

from fastapi import Depends, HTTPException, status

from app.core.deps import get_current_user
from app.models import Employee


# ─────────────────────────────────────────────────────────────────────────────
# Permission catalogue
# ─────────────────────────────────────────────────────────────────────────────
# Permissions are plain strings of the form ``<resource>:<action>``. The
# catalogue here is intentionally a superset of what the codebase currently
# enforces — routes can opt into the relevant permission without us having
# to define one ad hoc on the spot.

class Permission:
    # Employee management
    EMPLOYEE_READ            = "employee:read"
    EMPLOYEE_READ_ALL        = "employee:read_all"
    EMPLOYEE_UPDATE          = "employee:update"
    EMPLOYEE_UPDATE_SELF     = "employee:update_self"
    EMPLOYEE_CREATE          = "employee:create"
    EMPLOYEE_DELETE          = "employee:delete"

    # Leave management
    LEAVE_APPLY              = "leave:apply"
    LEAVE_VIEW_OWN           = "leave:view_own"
    LEAVE_VIEW_TEAM          = "leave:view_team"
    LEAVE_VIEW_ALL           = "leave:view_all"
    LEAVE_APPROVE_MANAGER    = "leave:approve_manager"
    LEAVE_APPROVE_HR         = "leave:approve_hr"
    LEAVE_ADMIN              = "leave:admin"  # cleanup / scheduler / overrides

    # Attendance / regularization
    ATTENDANCE_PUNCH         = "attendance:punch"
    ATTENDANCE_VIEW_OWN      = "attendance:view_own"
    ATTENDANCE_VIEW_TEAM     = "attendance:view_team"
    ATTENDANCE_VIEW_ALL      = "attendance:view_all"
    REGULARIZATION_REQUEST   = "regularization:request"
    REGULARIZATION_APPROVE   = "regularization:approve"

    # Timesheet
    TIMESHEET_SUBMIT         = "timesheet:submit"
    TIMESHEET_REVIEW         = "timesheet:review"
    TIMESHEET_PAYROLL_SYNC   = "timesheet:payroll_sync"

    # Onboarding
    CANDIDATE_MANAGE         = "candidate:manage"
    OFFER_MANAGE             = "offer:manage"
    BGV_INITIATE             = "bgv:initiate"
    BGV_VIEW                 = "bgv:view"
    CONVERT_CANDIDATE        = "candidate:convert"
    ACTIVATE_EMPLOYEE        = "employee:activate"

    # Notifications / search / admin tools
    NOTIFICATION_READ_OWN    = "notification:read_own"
    SEARCH_GLOBAL            = "search:global"
    ADMIN_ALL                = "admin:*"


# ─────────────────────────────────────────────────────────────────────────────
# Role → permission map
# ─────────────────────────────────────────────────────────────────────────────
# Mirrors the *effective* role behaviour that already exists across the
# codebase. To onboard a new role (e.g. ``it_admin``, ``hr``), add a row
# here and the entire codebase that uses :func:`require_permission` adapts.

ROLE_PERMISSIONS: Dict[str, Set[str]] = {
    "admin": {
        Permission.ADMIN_ALL,
        Permission.EMPLOYEE_READ_ALL, Permission.EMPLOYEE_READ,
        Permission.EMPLOYEE_CREATE, Permission.EMPLOYEE_UPDATE,
        Permission.EMPLOYEE_UPDATE_SELF, Permission.EMPLOYEE_DELETE,
        Permission.LEAVE_APPLY, Permission.LEAVE_VIEW_OWN, Permission.LEAVE_VIEW_TEAM,
        Permission.LEAVE_VIEW_ALL, Permission.LEAVE_APPROVE_MANAGER,
        Permission.LEAVE_APPROVE_HR, Permission.LEAVE_ADMIN,
        Permission.ATTENDANCE_PUNCH, Permission.ATTENDANCE_VIEW_OWN,
        Permission.ATTENDANCE_VIEW_TEAM, Permission.ATTENDANCE_VIEW_ALL,
        Permission.REGULARIZATION_REQUEST, Permission.REGULARIZATION_APPROVE,
        Permission.TIMESHEET_SUBMIT, Permission.TIMESHEET_REVIEW,
        Permission.TIMESHEET_PAYROLL_SYNC,
        Permission.CANDIDATE_MANAGE, Permission.OFFER_MANAGE,
        Permission.BGV_INITIATE, Permission.BGV_VIEW,
        Permission.CONVERT_CANDIDATE, Permission.ACTIVATE_EMPLOYEE,
        Permission.NOTIFICATION_READ_OWN, Permission.SEARCH_GLOBAL,
    },
    # ``hr`` is a synonym for the historical HR-flavoured admin used by
    # parts of the onboarding flow. Granting it the same permission set as
    # ``admin`` keeps existing behaviour when a JWT claims role=hr.
    "hr": {
        Permission.EMPLOYEE_READ_ALL, Permission.EMPLOYEE_READ,
        Permission.EMPLOYEE_CREATE, Permission.EMPLOYEE_UPDATE,
        Permission.EMPLOYEE_UPDATE_SELF,
        Permission.LEAVE_APPROVE_HR, Permission.LEAVE_VIEW_ALL,
        Permission.LEAVE_VIEW_TEAM, Permission.LEAVE_VIEW_OWN,
        Permission.ATTENDANCE_VIEW_ALL, Permission.ATTENDANCE_VIEW_TEAM,
        Permission.REGULARIZATION_APPROVE,
        Permission.CANDIDATE_MANAGE, Permission.OFFER_MANAGE,
        Permission.BGV_INITIATE, Permission.BGV_VIEW,
        Permission.CONVERT_CANDIDATE, Permission.ACTIVATE_EMPLOYEE,
        Permission.NOTIFICATION_READ_OWN,
    },
    "manager": {
        Permission.EMPLOYEE_READ, Permission.EMPLOYEE_UPDATE_SELF,
        Permission.LEAVE_APPLY, Permission.LEAVE_VIEW_OWN,
        Permission.LEAVE_VIEW_TEAM, Permission.LEAVE_APPROVE_MANAGER,
        Permission.ATTENDANCE_PUNCH, Permission.ATTENDANCE_VIEW_OWN,
        Permission.ATTENDANCE_VIEW_TEAM,
        Permission.REGULARIZATION_REQUEST, Permission.REGULARIZATION_APPROVE,
        Permission.TIMESHEET_SUBMIT, Permission.TIMESHEET_REVIEW,
        Permission.NOTIFICATION_READ_OWN, Permission.SEARCH_GLOBAL,
    },
    "employee": {
        Permission.EMPLOYEE_UPDATE_SELF, Permission.EMPLOYEE_READ,
        Permission.LEAVE_APPLY, Permission.LEAVE_VIEW_OWN,
        Permission.ATTENDANCE_PUNCH, Permission.ATTENDANCE_VIEW_OWN,
        Permission.REGULARIZATION_REQUEST,
        Permission.TIMESHEET_SUBMIT,
        Permission.NOTIFICATION_READ_OWN, Permission.SEARCH_GLOBAL,
    },
    # ``candidate`` users live in the legacy ``users`` table and only ever
    # talk to the candidate portal. They have no Core HRMS permissions.
    "candidate": set(),
    # Forward-compatibility placeholder. An IT admin can manage employees
    # (issue official email + temporary password) without leave/approval
    # power. Add real call sites as the role is rolled out.
    "it_admin": {
        Permission.EMPLOYEE_READ_ALL, Permission.EMPLOYEE_READ,
        Permission.EMPLOYEE_UPDATE, Permission.ACTIVATE_EMPLOYEE,
        Permission.NOTIFICATION_READ_OWN,
    },
}


def _normalise(role: str | None) -> str:
    return (role or "").strip().lower()


def permissions_for_role(role: str | None) -> Set[str]:
    """Return the effective permission set for a role name. Unknown roles
    receive an empty set (deny by default)."""
    return ROLE_PERMISSIONS.get(_normalise(role), set())


def has_permission(role: str | None, perm: str) -> bool:
    """Pure-Python check used by both the dependency and by ad-hoc call
    sites that have a role string but not a request context."""
    perms = permissions_for_role(role)
    if Permission.ADMIN_ALL in perms:
        return True
    return perm in perms


def require_permission(*required: str):
    """FastAPI dependency factory. Use exactly like ``role_required(...)``.

    Example::

        @router.post(
            "/leave/{id}/approve",
            dependencies=[Depends(require_permission(Permission.LEAVE_APPROVE_MANAGER))],
        )
        def approve_leave(...): ...

    The dependency reuses :func:`get_current_user`, which means existing
    JWT decoding + ``employment_status`` checks continue to apply — this
    is purely an authorization layer on top of the existing auth flow.
    """
    required_set = set(required)

    def dep(current: Employee = Depends(get_current_user)) -> Employee:
        role_name = current.role.name if current.role else None
        perms = permissions_for_role(role_name)
        if Permission.ADMIN_ALL in perms:
            return current
        if required_set.issubset(perms):
            return current
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail=f"Forbidden: missing permission(s) {sorted(required_set - perms)}",
        )

    return dep


__all__ = [
    "Permission",
    "ROLE_PERMISSIONS",
    "has_permission",
    "permissions_for_role",
    "require_permission",
]

File: Frontend/src/utils/permissions.js
// Centralised permission registry — frontend mirror of
// Backend/app/core/permissions.py.
//
// This module is **additive** and **opt-in**: every existing role check in
// the UI (`user.role === 'admin'`, etc.) continues to work. New code (and
// future refactors) can use `hasPermission(user, PERMISSIONS.X)` so that
// adding a new role only requires updating the table in one place.
//
// Keep this map in sync with the backend. The backend is the source of
// truth — frontend checks are UX hints only; the API still enforces
// authorization on every request.

export const PERMISSIONS = Object.freeze({
  // Employee management
  EMPLOYEE_READ:           'employee:read',
  EMPLOYEE_READ_ALL:       'employee:read_all',
  EMPLOYEE_UPDATE:         'employee:update',
  EMPLOYEE_UPDATE_SELF:    'employee:update_self',
  EMPLOYEE_CREATE:         'employee:create',
  EMPLOYEE_DELETE:         'employee:delete',

  // Leave management
  LEAVE_APPLY:             'leave:apply',
  LEAVE_VIEW_OWN:          'leave:view_own',
  LEAVE_VIEW_TEAM:         'leave:view_team',
  LEAVE_VIEW_ALL:          'leave:view_all',
  LEAVE_APPROVE_MANAGER:   'leave:approve_manager',
  LEAVE_APPROVE_HR:        'leave:approve_hr',
  LEAVE_ADMIN:             'leave:admin',

  // Attendance
  ATTENDANCE_PUNCH:        'attendance:punch',
  ATTENDANCE_VIEW_OWN:     'attendance:view_own',
  ATTENDANCE_VIEW_TEAM:    'attendance:view_team',
  ATTENDANCE_VIEW_ALL:     'attendance:view_all',
  REGULARIZATION_REQUEST:  'regularization:request',
  REGULARIZATION_APPROVE:  'regularization:approve',

  // Timesheet
  TIMESHEET_SUBMIT:        'timesheet:submit',
  TIMESHEET_REVIEW:        'timesheet:review',
  TIMESHEET_PAYROLL_SYNC:  'timesheet:payroll_sync',

  // Onboarding
  CANDIDATE_MANAGE:        'candidate:manage',
  OFFER_MANAGE:            'offer:manage',
  BGV_INITIATE:            'bgv:initiate',
  BGV_VIEW:                'bgv:view',
  CONVERT_CANDIDATE:       'candidate:convert',
  ACTIVATE_EMPLOYEE:       'employee:activate',

  // Misc
  NOTIFICATION_READ_OWN:   'notification:read_own',
  SEARCH_GLOBAL:           'search:global',
  ADMIN_ALL:               'admin:*',
});

const P = PERMISSIONS;

const ROLE_PERMISSIONS = Object.freeze({
  admin: new Set([
    P.ADMIN_ALL,
    P.EMPLOYEE_READ_ALL, P.EMPLOYEE_READ, P.EMPLOYEE_CREATE, P.EMPLOYEE_UPDATE,
    P.EMPLOYEE_UPDATE_SELF, P.EMPLOYEE_DELETE,
    P.LEAVE_APPLY, P.LEAVE_VIEW_OWN, P.LEAVE_VIEW_TEAM, P.LEAVE_VIEW_ALL,
    P.LEAVE_APPROVE_MANAGER, P.LEAVE_APPROVE_HR, P.LEAVE_ADMIN,
    P.ATTENDANCE_PUNCH, P.ATTENDANCE_VIEW_OWN, P.ATTENDANCE_VIEW_TEAM, P.ATTENDANCE_VIEW_ALL,
    P.REGULARIZATION_REQUEST, P.REGULARIZATION_APPROVE,
    P.TIMESHEET_SUBMIT, P.TIMESHEET_REVIEW, P.TIMESHEET_PAYROLL_SYNC,
    P.CANDIDATE_MANAGE, P.OFFER_MANAGE, P.BGV_INITIATE, P.BGV_VIEW,
    P.CONVERT_CANDIDATE, P.ACTIVATE_EMPLOYEE,
    P.NOTIFICATION_READ_OWN, P.SEARCH_GLOBAL,
  ]),
  hr: new Set([
    P.EMPLOYEE_READ_ALL, P.EMPLOYEE_READ, P.EMPLOYEE_CREATE, P.EMPLOYEE_UPDATE,
    P.EMPLOYEE_UPDATE_SELF,
    P.LEAVE_APPROVE_HR, P.LEAVE_VIEW_ALL, P.LEAVE_VIEW_TEAM, P.LEAVE_VIEW_OWN,
    P.ATTENDANCE_VIEW_ALL, P.ATTENDANCE_VIEW_TEAM, P.REGULARIZATION_APPROVE,
    P.CANDIDATE_MANAGE, P.OFFER_MANAGE, P.BGV_INITIATE, P.BGV_VIEW,
    P.CONVERT_CANDIDATE, P.ACTIVATE_EMPLOYEE, P.NOTIFICATION_READ_OWN,
  ]),
  manager: new Set([
    P.EMPLOYEE_READ, P.EMPLOYEE_UPDATE_SELF,
    P.LEAVE_APPLY, P.LEAVE_VIEW_OWN, P.LEAVE_VIEW_TEAM, P.LEAVE_APPROVE_MANAGER,
    P.ATTENDANCE_PUNCH, P.ATTENDANCE_VIEW_OWN, P.ATTENDANCE_VIEW_TEAM,
    P.REGULARIZATION_REQUEST, P.REGULARIZATION_APPROVE,
    P.TIMESHEET_SUBMIT, P.TIMESHEET_REVIEW,
    P.NOTIFICATION_READ_OWN, P.SEARCH_GLOBAL,
  ]),
  employee: new Set([
    P.EMPLOYEE_UPDATE_SELF, P.EMPLOYEE_READ,
    P.LEAVE_APPLY, P.LEAVE_VIEW_OWN,
    P.ATTENDANCE_PUNCH, P.ATTENDANCE_VIEW_OWN,
    P.REGULARIZATION_REQUEST,
    P.TIMESHEET_SUBMIT,
    P.NOTIFICATION_READ_OWN, P.SEARCH_GLOBAL,
  ]),
  candidate: new Set(),
  it_admin: new Set([
    P.EMPLOYEE_READ_ALL, P.EMPLOYEE_READ, P.EMPLOYEE_UPDATE, P.ACTIVATE_EMPLOYEE,
    P.NOTIFICATION_READ_OWN,
  ]),
});

function normaliseRole(role) {
  return (role || '').toString().trim().toLowerCase();
}

/** Returns the Set of effective permissions for a role name. */
export function permissionsForRole(role) {
  return ROLE_PERMISSIONS[normaliseRole(role)] || new Set();
}

/**
 * Pure check: does this user / role have a permission?
 * Accepts either a user object (`{ role }`) or a role string.
 */
export function hasPermission(userOrRole, perm) {
  const role = typeof userOrRole === 'string'
    ? userOrRole
    : (userOrRole && userOrRole.role);
  const perms = permissionsForRole(role);
  if (perms.has(PERMISSIONS.ADMIN_ALL)) return true;
  return perms.has(perm);
}

/** Returns true iff the user has ANY of the listed permissions. */
export function hasAnyPermission(userOrRole, perms) {
  return (perms || []).some((p) => hasPermission(userOrRole, p));
}

/** Returns true iff the user has ALL of the listed permissions. */
export function hasAllPermissions(userOrRole, perms) {
  return (perms || []).every((p) => hasPermission(userOrRole, p));
}
